### PR TITLE
refactor: consolidate `exitPosition` into `liquidatePosition`

### PR DESCRIPTION
**Pull request by fubuloubu**
fixes: #40 
fixes: #82

This PR consolidates the two similar functions `exitPosition` and `liquidatePosition` together, by adding another value that `liquidatePosition` returns (e.g. "realized loss"). The realized loss is reported during `Vault.report` (with `Strategy.emergencyShutdown` active) or `Vault.withdraw` and assessed to maintain the integrity of the Vault's balance sheet during those events. Further, the realized loss during withdrawals is fully applied to the withdrawer instead of socialized to the Vault, which should help prevent any "slippage" attacks against strategies connected to the Vault from being exploitable in practice.

### CODE REVIEW

File: contracts/BaseStrategy.sol [hunk 1]
```diff
@@ -432,23 +432,14 @@ abstract contract BaseStrategy {
...
-            uint256 _profit,
-            uint256 _loss,
-            uint256 _debtPayment
-        );
+     * Liquidate as many assets as possible to `want`, irregardless of slippage,
+     * up to `_amountNeeded`. Any excess will be re-invested with `adjustPosition()`.
+     * This function should return any realized losses incurred, and should also
+     * return the amount of `want` tokens made available by the liquidation.
+     * This function is used during emergency exit instead of `prepareReturn()` to
+     * liquidate all of the Strategy's positions back to the Vault.
```
Comment (Macarse) [inline contracts/BaseStrategy.sol:440]: liquidate `_amountNeeded` of ...

File: contracts/BaseStrategy.sol [hunk 2]
```diff
@@ -584,20 +575,27 @@ abstract contract BaseStrategy {
...
         if (emergencyExit) {
             // Free up as much capital as possible
-            // NOTE: Don't take performance fee in this scenario
-            (profit, loss, debtPayment) = exitPosition(vault.debtOutstanding());
+            uint256 totalAssets = estimatedTotalAssets();
+            // NOTE: use the larger of total assets or debt oustanding to get profit/loss calculation correct
+            (debtPayment, loss) = liquidatePosition(debtOutstanding > totalAssets ? debtOutstanding : totalAssets);
+            // NOTE: take up any remainder here as profit
+            if (debtPayment > debtOutstanding) {
+                profit = debtPayment.sub(debtOutstanding);
```
Comment (Macarse) [inline contracts/BaseStrategy.sol:587]: I don't understand why you take care of profits here and not losses? 
Why not make `liquidatePosition()` only `returns (uint256 _amountFreed)`

File: contracts/BaseStrategy.sol [hunk 3]
```diff
@@ -432,23 +432,14 @@ abstract contract BaseStrategy {
...
-    function exitPosition(uint256 _debtOutstanding)
-        internal
-        virtual
-        returns (
-            uint256 _profit,
-            uint256 _loss,
-            uint256 _debtPayment
-        );
+     * Liquidate as many assets as possible to `want`, irregardless of slippage,
+     * up to `_amountNeeded`. Any excess will be re-invested with `adjustPosition()`.
```
Comment (fubuloubu) [inline contracts/BaseStrategy.sol:436]:
```suggestion
     * Liquidate up to `_amountNeeded` of `want` of this strategy's positions,
     * irregardless of slippage. Any excess will be re-invested with `adjustPosition()`.
```

File: contracts/BaseStrategy.sol (hunk 1 above)
Comment (fubuloubu) [reply contracts/BaseStrategy.sol:440]: Is this clearer?
https://github.com/iearn-finance/yearn-vaults/pull/140/files#r549917601

File: contracts/BaseStrategy.sol (hunk 2 above)
Comment (fubuloubu) [reply contracts/BaseStrategy.sol:587]: The assumption is that `liquidatePoisition` should be used to make the Vault whole as quickly as possible, up to `_amountNeeded`. Here we are sending in the maximum amount of assets possible for the Strategy to liquidate.

If there is anything left after paying back the debt, the rest is booked as profits.

Perhaps this is over complicated though? My intention was to use this function in both the `withdraw` and `emergencyShutdown` scenario, because under that scenario we want to withdraw the assets as quickly as possible, slippage be damned.

File: contracts/BaseStrategy.sol (hunk 1 above)
Comment (Macarse) [reply contracts/BaseStrategy.sol:440]: yep!

File: contracts/BaseStrategy.sol (hunk 3 above)
Comment (Macarse) [reply contracts/BaseStrategy.sol:436]: merge

File: contracts/BaseStrategy.sol (hunk 2 above)
Comment (Macarse) [reply contracts/BaseStrategy.sol:587]: my point is if you have a liquidate position that just returns the amount freed, in `harvest()` you can do:

```
if (debtPayment > debtOutstanding) {
  // profit
} else if (debtPayment < debtOutstanding) {
  // loss
}

Comment (fubuloubu) [reply contracts/BaseStrategy.sol:587]: I think I see your point. Let me see how it looks

File: contracts/BaseStrategy.sol [hunk 4]
```diff
@@ -584,20 +577,28 @@ abstract contract BaseStrategy {
...
         uint256 profit = 0;
         uint256 loss = 0;
+        uint256 debtOutstanding = vault.debtOutstanding();
         uint256 debtPayment = 0;
         if (emergencyExit) {
             // Free up as much capital as possible
-            // NOTE: Don't take performance fee in this scenario
-            (profit, loss, debtPayment) = exitPosition(vault.debtOutstanding());
+            uint256 totalAssets = estimatedTotalAssets();
+            debtPayment = liquidatePosition(totalAssets);
```
Comment (Macarse) [inline contracts/BaseStrategy.sol:585]: do you mind using a different variable to make it clear? something like `liquidatedAmount`?

**Comment by fubuloubu** [comment]
Consider adding an extra default parameter to `Vault.withdraw` that allows the withdrawer to specify a maximum amount of slippage loss that is acceptable (and revert above that)

**Comment by fubuloubu** [comment]
Note: add back loss reporting in `liquidatePosition` for corner case when strategy has "locked" funds

File: contracts/BaseStrategy.sol [hunk 5]
```diff
@@ -432,23 +432,17 @@ abstract contract BaseStrategy {
...
-        );
+     * Liquidate up to `_amountNeeded` of `want` of this strategy's positions,
+     * irregardless of slippage. Any excess will be re-invested with `adjustPosition()`.
+     * This function should return the amount of `want` tokens made available by the
+     * liquidation. If there is a difference between them, `_loss` indicates whether the
+     * difference is due to a realized loss, or if there is some other sitution at play
+     * (e.g. locked funds). This function is used during emergency exit instead of
+     * `prepareReturn()` to liquidate all of the Strategy's positions back to the Vault.
+     *
+     * NOTE: The invariant `_amountFreed <= _amountNeeded` should always be maintained
```
Comment (fubuloubu) [inline contracts/BaseStrategy.sol:443]:
```suggestion
     * NOTE: The invariant `_amountFreed + _loss <= _amountNeeded` should always be maintained
```

**Review by Macarse — APPROVED** [review]

File: contracts/BaseStrategy.sol [hunk 6]
```diff
@@ -608,26 +609,23 @@ abstract contract BaseStrategy {
...
+    function withdraw(uint256 _amountNeeded) external returns (uint256 loss) {
         require(msg.sender == address(vault), "!vault");
         // Liquidate as much as possible to `want`, up to `_amount`
-        uint256 amountFreed = liquidatePosition(_amountNeeded);
+        uint256 amountFreed;
+        (amountFreed, loss) = liquidatePosition(_amountNeeded);
         // Send it directly back (NOTE: Using `msg.sender` saves some gas here)
         want.transfer(msg.sender, amountFreed);
+        // Reinvest anything leftover
+        adjustPosition(vault.debtOutstanding());
```
Comment (dmolina79) [inline contracts/BaseStrategy.sol:628]: should this be a concern of tend instead of withdraw to reinvest leftover?

File: contracts/BaseStrategy.sol [hunk 7]
```diff
@@ -608,26 +609,23 @@ abstract contract BaseStrategy {
...
      */
-    function withdraw(uint256 _amountNeeded) external {
+    function withdraw(uint256 _amountNeeded) external returns (uint256 loss) {
         require(msg.sender == address(vault), "!vault");
         // Liquidate as much as possible to `want`, up to `_amount`
-        uint256 amountFreed = liquidatePosition(_amountNeeded);
+        uint256 amountFreed;
+        (amountFreed, loss) = liquidatePosition(_amountNeeded);
         // Send it directly back (NOTE: Using `msg.sender` saves some gas here)
         want.transfer(msg.sender, amountFreed);
```
Comment (dmolina79) [inline contracts/BaseStrategy.sol:626]: should we do instead
`want.transfer(msg.sender, Min(amountNeeded, amountFreed))` 

for the transfer, in case some strategy borks the accounting?

Comment (fubuloubu) [reply contracts/BaseStrategy.sol:626]: No, I would say that's a bad strategy. See
https://github.com/iearn-finance/yearn-vaults/pull/140/files#diff-517cca2fbf046cbfefd26edb1fec816e5b2a5571fd99927dc81d6b4be60cd2a7R443

Comment (fubuloubu) [reply contracts/BaseStrategy.sol:626]: I guess I could add an assertion for this though...

File: contracts/BaseStrategy.sol (hunk 6 above)
Comment (fubuloubu) [reply contracts/BaseStrategy.sol:628]: good point, this could also be abused as an attack vector to force a strategy to rebalance itself, potentially into a bad position.

File: contracts/BaseStrategy.sol (hunk 7 above)
Comment (dmolina79) [reply contracts/BaseStrategy.sol:626]: yeah, i was thinking of how to mitigate if the invariant is not maintained, and i don't like adding a revert or require check also since it may lock up stuff, but i guess sending more than needed is still not the end of the world since withdraw on vault also checks for amount of shares, those "additional" funds just are transferred to the vault i guess.

Comment (fubuloubu) [reply contracts/BaseStrategy.sol:626]: This is exactly right, looks like it is protected! Won't add the require statement.

**Review by dmolina79 — APPROVED** [review]
LGTM!

File: contracts/BaseStrategy.sol [hunk 8]
```diff
@@ -432,23 +432,17 @@ abstract contract BaseStrategy {
...
-        );
+     * Liquidate up to `_amountNeeded` of `want` of this strategy's positions,
+     * irregardless of slippage. Any excess will be re-invested with `adjustPosition()`.
+     * This function should return the amount of `want` tokens made available by the
+     * liquidation. If there is a difference between them, `_loss` indicates whether the
+     * difference is due to a realized loss, or if there is some other sitution at play
+     * (e.g. locked funds). This function is used during emergency exit instead of
+     * `prepareReturn()` to liquidate all of the Strategy's positions back to the Vault.
+     *
+     * NOTE: The invariant `_amountFreed + _loss <= _amountNeeded` should always be maintained
```
Comment (mushroomsforest) [inline contracts/BaseStrategy.sol:443]: the `_amountFreed` is same as `_liquidatedAmount` ?
